FIX: Add run requirement for pythia8 variant version

matthewfeickert opened the pull request:
* Add `run` requirement of `pythia8 ${{ pythia8 }}.*` to restrict `pythia8` to
  the variant version used as a `host` requirement during the build for ABI stability.
   - This needs to be a `run` requirement and not a `run_export` as `run_export`s are weak requirements and a strong requirement is needed to override `pythia8`'s existing `run_export`s.
* Unrestrict `pythia8` variants for `linux-aarch64` to get `v8.312` based builds.
   - c.f https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
* Bump build number.

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [N/A] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

matthewfeickert committed:
FIX: Add run requirement for pythia8 variant version

* Add 'run' requirement of 'pythia8 ${{ pythia8 }}.*' to restrict pythia8 to
  the variant version used as a 'host' requirement during the build for ABI
  stability.
   - This needs to be a 'run' requirement and not a 'run_export' as run_exports
     are weak requirements and a strong requirement is needed to override
     pythia8's existing run_exports.
* Unrestrict pythia8 variants for linux-aarch64 to get v8.312 based builds.
* Bump build number.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.

matthewfeickert commented:
With

```yaml
  host:
    - hepmc2
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 ==8.311
      else: pythia8
```

the build gives

```
 │ │ Finalized run dependencies (mg5amcnlo-pythia8-interface-1.3.0-hbd9a09b_3):
 │ │ ╭──────────────────┬────────────────────────────────────────────╮
 │ │ │ Name             ┆ Spec                                       │
 │ │ ╞══════════════════╪════════════════════════════════════════════╡
 │ │ │ Run dependencies ┆                                            │
 │ │ │ libstdcxx        ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ libgcc           ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ pythia8          ┆ >=8.311,<8.400.0a0 (RE of [host: pythia8]) │
 │ │ ╰──────────────────┴────────────────────────────────────────────╯
```

so I guess explicit `run` requirements are needed(?).

matthewfeickert commented:
And

```yaml
  host:
    - hepmc2
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 8.311.*
      else: pythia8
  run:
    - hepmc2
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 8.311.*
      else: pythia8
```

gives

```
 │ │ Finalized run dependencies (mg5amcnlo-pythia8-interface-1.3.0-hbd9a09b_3):
 │ │ ╭──────────────────┬────────────────────────────────────────────╮
 │ │ │ Name             ┆ Spec                                       │
 │ │ ╞══════════════════╪════════════════════════════════════════════╡
 │ │ │ Run dependencies ┆                                            │
 │ │ │ hepmc2           ┆                                            │
 │ │ │ pythia8          ┆ 8.311.*                                    │
 │ │ │ libstdcxx        ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ libgcc           ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ pythia8          ┆ >=8.311,<8.400.0a0 (RE of [host: pythia8]) │
 │ │ ╰──────────────────┴────────────────────────────────────────────╯
```

matthewfeickert commented:
And

```yaml
  host:
    - hepmc2
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 ==8.311
      else: pythia8
  run:
    - hepmc2
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 ==8.311
      else: pythia8
```

gives

```
 │ │ Run dependencies:
 │ │ ╭──────────────────┬────────────────────────────────────────────╮
 │ │ │ Name             ┆ Spec                                       │
 │ │ ╞══════════════════╪════════════════════════════════════════════╡
 │ │ │ Run dependencies ┆                                            │
 │ │ │ hepmc2           ┆                                            │
 │ │ │ pythia8          ┆ ==8.311                                    │
 │ │ │ libstdcxx        ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ libgcc           ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ pythia8          ┆ >=8.311,<8.400.0a0 (RE of [host: pythia8]) │
 │ │ ╰──────────────────┴────────────────────────────────────────────╯
```

So I guess for `run` requirements that `8.311.*` and `==8.311` result in basically the same dependencies.

matthewfeickert commented:
I think the problem is that [`pythia8` has a `run_exports`](https://github.com/conda-forge/pythia8-feedstock/blob/ba82aff4b9c1154568747381cd4770acf1451f9a/recipe/meta.yaml#L22-L24) of

```yaml
  run_exports:
    # https://abi-laboratory.pro/?view=timeline&l=root
    - pythia8 >={{ version }},<{{ version.split('.')[0] }}.{{ (version.split('.')[1][0]|int) + 1 }}00.0a0
    # Can't use pin_subpackage as X.Y.Z is written as X.YZ
```

which gives the

```
 │ │ ╭──────────────────┬────────────────────────────────────────────╮
 │ │ │ Name             ┆ Spec                                       │
 │ │ ╞══════════════════╪════════════════════════════════════════════╡
 │ │ │ Run dependencies ┆                                            │
...
 │ │ │ pythia8          ┆ >=8.311,<8.400.0a0 (RE of [host: pythia8]) │
 │ │ ╰──────────────────┴────────────────────────────────────────────╯
```

and so `mg5amcnlo-pythia8-interface` needs to define `run` requirements or `run_exports` of its own as well.

matthewfeickert commented:
For `linux-aarch64`

```yaml
  host:
    - hepmc2
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 ==8.311
      else: pythia8
  run_exports:
    - hepmc2
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 ==8.311
      else: pythia8
```

gives

```
 │ │ Finalized run dependencies (mg5amcnlo-pythia8-interface-1.3.0-hbd9a09b_3):
 │ │ ╭────────────────────┬────────────────────────────────────────────╮
 │ │ │ Name               ┆ Spec                                       │
 │ │ ╞════════════════════╪════════════════════════════════════════════╡
 │ │ │ Run dependencies   ┆                                            │
 │ │ │ libstdcxx          ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ libgcc             ┆ >=14 (RE of [build: gxx_linux-aarch64])    │
 │ │ │ pythia8            ┆ >=8.311,<8.400.0a0 (RE of [host: pythia8]) │
 │ │ │                    ┆                                            │
 │ │ │ Run exports (Weak) ┆                                            │
 │ │ │ hepmc2             ┆                                            │
 │ │ │ pythia8            ┆ ==8.311                                    │
 │ │ ╰────────────────────┴────────────────────────────────────────────╯
```

But I'm seeing that for `linux-64` that the run exports from `pythia8` disallow being able to optionally define `pythia8` `v8.311` if a user wanted to

```
 │ │ Run dependencies:
 │ │ ╭────────────────────┬──────────────────────────────────────────────────╮
 │ │ │ Name               ┆ Spec                                             │
 │ │ ╞════════════════════╪══════════════════════════════════════════════════╡
 │ │ │ Run dependencies   ┆                                                  │
 │ │ │ libstdcxx          ┆ >=14 (RE of [build: gxx_linux-64])               │
 │ │ │ libgcc             ┆ >=14 (RE of [build: gxx_linux-64])               │
 │ │ │ __glibc            ┆ >=2.17,<3.0.a0 (RE of [build: sysroot_linux-64]) │
 │ │ │ pythia8            ┆ >=8.312,<8.400.0a0 (RE of [host: pythia8])       │
 │ │ │                    ┆                                                  │
 │ │ │ Run exports (Weak) ┆                                                  │
 │ │ │ hepmc2             ┆                                                  │
 │ │ │ pythia8            ┆                                                  │
 │ │ ╰────────────────────┴──────────────────────────────────────────────────╯
```

```console
# pixi init example && cd example && pixi add mg5amcnlo-pythia8-interface && pixi list pythia
Created /example/pixi.toml
Added mg5amcnlo-pythia8-interface >=1.3.0,<2
Package                      Version  Build              Size       Kind   Source
mg5amcnlo-pythia8-interface  1.3.0    h9e02651_2         137.9 KiB  conda  https://conda.anaconda.org/conda-forge/
pythia8                      8.312    py313h6f690f4_104  20.7 MiB   conda  https://conda.anaconda.org/conda-forge/
# pixi add mg5amcnlo-pythia8-interface 'pythia8 8.311'
Added mg5amcnlo-pythia8-interface >=1.3.0,<2
Added pythia8 8.311
# pixi list pythia
Package                      Version  Build              Size       Kind   Source
mg5amcnlo-pythia8-interface  1.3.0    h48bac2b_1         138.5 KiB  conda  https://conda.anaconda.org/conda-forge/
pythia8                      8.311    py313h4fcea12_108  21.1 MiB   conda  https://conda.anaconda.org/conda-forge/
```

Note how the `mg5amcnlo-pythia8-interface` build number regressed back to `1` when `pythia8` `v8.311` was requested.

```console
# pixi add mg5amcnlo-pythia8-interface[build_number=2] 'pythia8 8.311'
Error:   x failed to solve requirements of environment 'default' for platform 'linux-64'
  |->   x failed to solve the environment
  |   
  `-> Cannot solve the request because of: mg5amcnlo-pythia8-interface * cannot be installed because there are no viable options:
      └─ mg5amcnlo-pythia8-interface 1.3.0 would require
         └─ pythia8 >=8.312,<8.400.0a0, for which no candidates were found.
      

```

I think(?) this means that to support both `pythia8` `v8.311` and `v8.312` that I need to use variants. I also need to figure out how to disallow a variant for a platform (as `pythia8` `v8.312` is currently not viable for `linux-aarch64`).

matthewfeickert commented:
**edit**: Never mind. `linux-aarch64` builds of `mg5amcnlo-pythia8-interface` and `pythia8` `v8.312` are broken given testing in https://github.com/conda-forge/mg5amcnlo-feedstock/pull/8.

[linux-aarch64-pythia8-8.312-build-failure.txt](https://github.com/user-attachments/files/22589623/linux-aarch64-pythia8-8.312-build-failure.txt)

~~It might not be that `linux-aarch64` is broken, but that the version of `mg5amcnlo` that can work with `pythia8` is `build` `2` (which should probably be marked as broken, but I have been waiting until `mg5amcnlo` is actually working for NLO processes across all modern platforms to go back and mark all previous builds as broken) and the current build is `build` `5`~~

```console
$ pixi init example && cd example && pixi add mg5amcnlo 'pythia8 8.312' && pixi list mg5
Created /example/pixi.toml
Added mg5amcnlo >=3.5.7,<4
Added pythia8 8.312
Package    Version  Build            Size      Kind   Source
mg5amcnlo  3.5.7    py311hd817590_2  15.9 MiB  conda  https://conda.anaconda.org/conda-forge/
$ pixi clean && pixi add mg5amcnlo 'pythia8 8.311' && pixi list mg5
Added mg5amcnlo >=3.5.7,<4
Added pythia8 8.311
Package    Version  Build            Size      Kind   Source
mg5amcnlo  3.5.7    py311h610b25d_5  19.6 MiB  conda  https://conda.anaconda.org/conda-forge/
```

~~So to properly test this, I think https://github.com/conda-forge/admin-requests/pull/1675 will need to get merged to mark `mg5amcnlo-pythia8-interface` `v1.3.0` `build` `2` as broken, and then https://github.com/conda-forge/mg5amcnlo-feedstock/pull/8 will need to get rerun with `pythia8` `8.312` added as a dependency for debug purposes. If that passes then `linux-aarch64` `v8.312` is not broken.~~

Regardless, to get this PR fixed I'll need to determine how run exports for the version of `pythia8` that was used in `host` can work.

matthewfeickert commented:
I'm also confused as the [`conda-build` docs on `variants`](https://docs.conda.io/projects/conda-build/en/stable/resources/variants.html#pinning-at-the-variant-level) show

`recipe/conda_build_config.yaml`

```yaml
boost: 1.63
pin_run_as_build:
    boost:
      max_pin: x.x
```

`recipe/recipe.yaml`

```yaml
requirements:
    build:
        - boost
    run:
        - boost
```

should result in a 

> runtime boost dependency will be pinned to `>=(current boost 1.63.x version),<1.64`.

But for my recipe's `recipe/conda_build_config.yaml`

```yaml
pin_run_as_build:
  pythia8:
    max_pin: "x.x.x"  # Should pin down to the patch level

# c.f. https://github.com/conda-forge/staged-recipes/issues/27750 for why not
# currently using variants.yaml
pythia8:
- "8.311"
- "8.312"
```

and `recipe/recipe.yaml`

```yaml
requirements:
  build:
    - ${{ stdlib('c') }}
    - ${{ compiler('cxx') }}
    - make
    - if: build_platform != target_platform
      then:
        - python
  host:
    - hepmc2
    - pythia8  # variant
  run_exports:
    - hepmc2
  # Need to use 'run' instead of 'run_exports' as run_exports are weak
  # and will be overriden by pythia8
  run:
    # c.f. https://github.com/conda-forge/mg5amcnlo-pythia8-interface-feedstock/issues/2
    - if: linux and aarch64
      then: pythia8 ==8.311
      else: pythia8  # variant
```

I get (for example `linux-64`)

```
 │ │ Run dependencies:
 │ │ ╭────────────────────┬──────────────────────────────────────────────────╮
 │ │ │ Name               ┆ Spec                                             │
 │ │ ╞════════════════════╪══════════════════════════════════════════════════╡
 │ │ │ Run dependencies   ┆                                                  │
 │ │ │ pythia8            ┆                                                  │
 │ │ │ __glibc            ┆ >=2.17,<3.0.a0 (RE of [build: sysroot_linux-64]) │
 │ │ │ libstdcxx          ┆ >=14 (RE of [build: gxx_linux-64])               │
 │ │ │ libgcc             ┆ >=14 (RE of [build: gxx_linux-64])               │
 │ │ │ pythia8            ┆ >=8.311,<8.400.0a0 (RE of [host: pythia8])       │
 │ │ │                    ┆                                                  │
 │ │ │ Run exports (Weak) ┆                                                  │
 │ │ │ hepmc2             ┆                                                  │
 │ │ ╰────────────────────┴──────────────────────────────────────────────────╯
``` 

Why does the `max_pin` of `"x.x."` not pin `pythia8` to `>=8.311,<8.312.0a0`?

matthewfeickert commented:
Aside: Something that I've noticed that I think is related to the `linux-aarch64` segfaults is that in all the builds there are different objects being linked together where at least one of them was built with GCC 4.8.5 that is _not_ the conda-forge distribution:

```console
$ docker run --rm -ti ghcr.io/prefix-dev/pixi:latest 
root@0b7caefcc571:/# pixi init example && cd example && pixi add mg5amcnlo-pythia8-interface 'pythia8 8.311' cxx-compiler
Created /example/pixi.toml
Added mg5amcnlo-pythia8-interface >=1.3.0,<2
Added pythia8 8.311
Added cxx-compiler >=1.11.0,<2
root@0b7caefcc571:/example# pixi shell 

(example) root@0b7caefcc571:/example# command -v readelf
/example/.pixi/envs/default/bin/readelf
(example) root@0b7caefcc571:/example# readelf -p .comment $CONDA_PREFIX/MG5_aMC/HEPTools/MG5aMC_PY8_interface/MG5aMC_PY8_interface

String dump of section '.comment':
  [     0]  GCC: (GNU) 4.8.5 20150623 (Red Hat 4.8.5-39)
  [    2d]  GCC: (conda-forge gcc 14.3.0-5) 14.3.0

(example) root@0b7caefcc571:/example# pixi list mg5
Package                      Version  Build       Size       Kind   Source
mg5amcnlo-pythia8-interface  1.3.0    h48bac2b_1  138.5 KiB  conda  https://conda.anaconda.org/conda-forge/
```

and if I fire up a Raspberry Pi to get a `linux-aarch64` build

```console
$ screenfetch 
         _,met$$$$$gg.           pi@raspberrypi-64
      ,g$$$$$$$$$$$$$$$P.        OS: Debian 11 bullseye
    ,g$$P""       """Y$$.".      Kernel: aarch64 Linux 6.1.21-v8+
   ,$$P'              `$$$.      Uptime: 1h 48m
  ',$$P       ,ggs.     `$$b:    Packages: 1422
  `d$$'     ,$P"'   .    $$$     Shell: bash 5.2.37
   $$P      d$'     ,    $$P     Disk: 9.6G / 118G (9%)
   $$:      $$.   -    ,d$$'     CPU: BCM2835 @ 4x 1.5GHz
   $$\;      Y$b._   _,d$P'      RAM: 387MiB / 3794MiB
   Y$$.    `.`"Y$$$$P"'         
   `$$b      "-.__              
    `Y$$                        
     `Y$$.                      
       `$$b.                    
         `Y$$b.                 
            `"Y$b._             
                `""""           
                                
$ pixi init example && cd example && pixi add mg5amcnlo-pythia8-interface 'pythia8 8.311' cxx-compiler && pixi list mg5
✔ Created /tmp/example/pixi.toml
✔ Added mg5amcnlo-pythia8-interface >=1.3.0,<2
✔ Added pythia8 8.311
✔ Added cxx-compiler >=1.11.0,<2
Package                      Version  Build       Size       Kind   Source
mg5amcnlo-pythia8-interface  1.3.0    h4860862_1  168.2 KiB  conda  https://conda.anaconda.org/conda-forge/
pi@raspberrypi-64:/tmp/example $ pixi shell

(example) pi@raspberrypi-64:/tmp/example $ readelf -p .comment $CONDA_PREFIX/MG5_aMC/HEPTools/MG5aMC_PY8_interface/MG5aMC_PY8_interface

String dump of section '.comment':
  [     0]  GCC: (GNU) 4.8.5 20150623 (Red Hat 4.8.5-39)
  [    2d]  GCC: (conda-forge gcc 14.3.0-5) 14.3.0

(example) pi@raspberrypi-64:/tmp/example $ 
```

As the recipe uses

```yaml
- $CXX ./MG5aMC_PY8_interface.cc -o MG5aMC_PY8_interface $CXXFLAGS $(pythia8-config --cxxflags) $LDFLAGS $(pythia8-config --ldflags) -lHepMC
```

and

```console
(example) pi@raspberrypi-64:/tmp/example $ echo "$LDFLAGS $(pythia8-config --ldflags) -lHepMC"
-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--allow-shlib-undefined -Wl,-rpath,/tmp/example/.pixi/envs/default/lib -Wl,-rpath-link,/tmp/example/.pixi/envs/default/lib -L/tmp/example/.pixi/envs/default/lib -L/tmp/example/.pixi/envs/default/lib -Wl,-rpath,/tmp/example/.pixi/envs/default/lib -lpythia8 -L/tmp/example/.pixi/envs/default/lib -Wl,-rpath,/tmp/example/.pixi/envs/default/lib -lz -ldl -lHepMC
```

then looking at the linked libraries

```console
(example) pi@raspberrypi-64:/tmp/example $ readelf -p .comment $CONDA_PREFIX/lib/libpythia8.so

String dump of section '.comment':
  [     0]  GCC: (conda-forge gcc 14.3.0-4) 14.3.0

(example) pi@raspberrypi-64:/tmp/example $ readelf -p .comment $CONDA_PREFIX/lib/libz.so

String dump of section '.comment':
  [     0]  GCC: (conda-forge gcc 13.3.0-1) 13.3.0

(example) pi@raspberrypi-64:/tmp/example $ readelf -p .comment $CONDA_PREFIX/lib/libHepMC.so

String dump of section '.comment':
  [     0]  GCC: (conda-forge gcc 13.3.0-1) 13.3.0

(example) pi@raspberrypi-64:/tmp/example $ readlink -f $(find $CONDA_PREFIX/ -type f -iname "libc*.so")
/tmp/example/.pixi/envs/default/aarch64-conda-linux-gnu/sysroot/usr/lib64/libc.so
/tmp/example/.pixi/envs/default/aarch64-conda-linux-gnu/sysroot/usr/lib64/gconv/libCNS.so
/tmp/example/.pixi/envs/default/aarch64-conda-linux-gnu/sysroot/lib64/libc-2.28.so
(example) pi@raspberrypi-64:/tmp/example $ readelf -p .comment /tmp/example/.pixi/envs/default/aarch64-conda-linux-gnu/sysroot/lib64/libc-2.28.so

String dump of section '.comment':
  [     0]  GCC: (GNU) 8.5.0 20210514 (Red Hat 8.5.0-20)
```

So it seems the non-conda-forge compiler is coming from `ldl` which are coming from sysroot's `libc`. So is it possible that 

```console
$ pythia8-config --ldflags
-L/tmp/example/.pixi/envs/default/lib -Wl,-rpath,/tmp/example/.pixi/envs/default/lib -lpythia8 -L/tmp/example/.pixi/envs/default/lib -Wl,-rpath,/tmp/example/.pixi/envs/default/lib -lz -ldl
```

is causing this to go wrong because the `build` environment's sysroot's `libc`is very old?

**edit**: No. So I'm confused again.

```console
$ docker run --rm -ti -v $PWD:/read ghcr.io/prefix-dev/pixi:latest 
root@3a7cea45c1a8:/# pixi init example && cd example && pixi add cxx-compiler && pixi shell
Created /example/pixi.toml
Added cxx-compiler >=1.11.0,<2

(example) root@3a7cea45c1a8:/example# pixi global install vim
└── vim: 9.1.1627 (installed)
    └─ exposes: ex, rview, rvim, view, vim, vimdiff, vimtutor, xxd
(example) root@3a7cea45c1a8:/example# vim test.cpp
(example) root@3a7cea45c1a8:/example# cat test.cpp 
#include <iostream>

int main() {
    std::cout << "Hello, World!" << std::endl;

    return 0;
}

(example) root@3a7cea45c1a8:/example# $CXX ./test.cpp -o test
(example) root@3a7cea45c1a8:/example# ./test 
Hello, World!
(example) root@3a7cea45c1a8:/example# readelf -p .comment ./test

String dump of section '.comment':
  [     0]  GCC: (GNU) 8.5.0 20210514 (Red Hat 8.5.0-20)
  [    2d]  GCC: (conda-forge gcc 14.3.0-5) 14.3.0

(example) root@3a7cea45c1a8:/example# command -v $CXX
/example/.pixi/envs/default/bin/x86_64-conda-linux-gnu-c++
(example) root@3a7cea45c1a8:/example# readelf -p .comment $(command -v $CXX)

String dump of section '.comment':
  [     0]  GCC: (GNU) 4.8.5 20150623 (Red Hat 4.8.5-39)
  [    2d]  GCC: (conda-forge gcc 15.1.0-4) 15.1.0

```

matthewfeickert commented:
So the current `build` `1` of `mg5amcnlo-pythia8-interface`

```console
$ pixi list mg5
Package                      Version  Build       Size       Kind   Source
mg5amcnlo-pythia8-interface  1.3.0    h3cf2d66_1  138.1 KiB  conda  https://conda.anaconda.org/conda-forge/
```

is broken as the constraints on the variants were not correct during that build. What is required is that the version of `pythia8` that the `mg5amcnlo-pythia8-interface` build is built against is installed with it, which means that it is required to find some way to set the exact version of the `host` `pythia8` in the `run_exports`.

matthewfeickert commented:
Turns out that just pinning to the variant package version works

```yaml
requirements:
  build:
    - ${{ stdlib('c') }}
    - ${{ compiler('cxx') }}
    - make
    - if: build_platform != target_platform
      then:
        - python
  host:
    - hepmc2
    - pythia8  # variant
  run_exports:
    - hepmc2
  run:
    # Need to use 'run' instead of 'run_exports' for pythia8 as run_exports are
    # weak requirements, and need a strong requirement ('run') to override
    # pythia8's run_exports
    - pythia8 ${{ pythia8 }}.*  # pin to variant
```

and that [`rattler-build`'s `pin_run_as_build`](https://rattler.build/latest/variants/#pin-run-as-build) information isn't getting used at all, so it can be removed with no effect

```diff
diff --git a/recipe/variants.yaml b/recipe/variants.yaml
index 6fcd4a0..2b0f0fe 100644
--- a/recipe/variants.yaml
+++ b/recipe/variants.yaml
@@ -1,7 +1,3 @@
-pin_run_as_build:
-  pythia8:
-    max_pin: "x.x.x"
-
```

I don't understand why though, which is confusing.

matthewfeickert committed:
MNT: Re-rendered with conda-smithy 3.52.3 and conda-forge-pinning 2025.09.27.05.44.00

Other tools:
- conda-build 25.7.0
- rattler-build 0.47.0
- rattler-build-conda-compat 1.4.6
